src/Suave/Logging/LogLine.fs

```diff
@@ -13,23 +15,45 @@ type LogLine =
     level         : LogLevel
     /// the source of the log line, e.g. 'ModuleName.FunctionName'
     path          : string
-    /// the message that the application wants to log
-    message       : string
+    /// the template message that the application wants to log
+    template      : string
+    /// a function which can format the log template message. When the
+    /// message is pre-formatted, this is None. When the log message 
+    /// contains structured data, this function can be used to render
+    /// the message template with the data.
+    formatter     : LogTemplateFormatter option
+    /// any key-value based data to log
+    data          : Map<string, obj>
     /// an optional exception
     ``exception`` : exn option
     /// timestamp when this log line was created
     tsUTCTicks    : int64 }
-    
+
+    /// Formats the log line as a string
+    member x.formatMessage () =
+        match x.formatter with
+        | None -> x.template
+        | Some fmter -> fmter x.template x.GetDataByName
+
+    /// Gets a data value
+    member x.GetDataByName name = Option.toObj (x.data.TryFind name)
+
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module LogLine =
   let mk path level trace ex message =
-    { message       = message
+    { template      = message
+      formatter     = None
+      data          = Map.empty
       level         = level
       path          = path
       ``exception`` = ex
       trace         = trace
       tsUTCTicks  = Globals.utcNow().Ticks }
 
+  /// assists with type inference, avoiding ugly 'box' calls for each
+  /// value item when constructing the log data map 
+  let mapLogData (data: (string*obj) list) = Map.ofList data
```

Comment:
OK - there are key questions I still have, and I think I need your help.

MessageTemplates & Serilog takes the approach that the log message creator can provide any object that follows some basic rules for "destructuring". You would like to avoid the destructuring complexity (and associated reflection involved) and replace it with message template construction inconvenience (just how much of an inconvenience remains to be seen).

In the MessageTemplates/Serilog model, the log message creator would do something like this:
```fsharp
log.Information "listener started in {startData}ms" startData
```

In your proposed model (inc. Logary), the log message creator would do something like this:
```fsharp
log.Information("listener started in {???} ms", convertToStructure ("startData", startData))`
// additionally, some small amount of static member constraints that allow
// startData to be converted into your object model, this must be written manually.
```

In your proposed _structured log message_ model, you haven't yet solved the issue of how to store the string template messages, and how to parse and match them to the provided arguments. Is that correct, or have I missed something?

I don't yet understand how the 'lens' complexity is justified here, could you explain or point me to something that might help clear this up for me?